x/ecocredit/spec/06_tests.md

```diff
@@ -281,3 +269,185 @@ If a user tries to cancel credits and the user provides an invalid batch denomin
+
+- GIVEN - user provides a bid price denomination that is NOT a match of the sell order ask price denomination
+- WHEN - user tries to create a buy order
+- THEN - transaction fails, buy order is NOT processed
+
+If a user tries to create a buy order and the user provides a quantity that amounts to a total cost that is less than or equal to their coin balance, then the transaction is successful and the buy order is processed.
+
+- GIVEN - user provides a quantity that amounts to a total cost that is less than or equal to their coin balance
+- WHEN - user tries to create a buy order
+- THEN - transaction is successful, buy order is processed
+
+If a user tries to create a buy order and the user provides a quantity that amounts to a total cost that is more than their coin balance, then the transaction fails and the buy order is NOT processed.
+
+- GIVEN - user provides a quantity that amounts to a total cost that is more than their coin balance
+- WHEN - user tries to create a buy order
+- THEN - transaction fails, buy order is NOT processed
+
+If a user tries to create a buy order and the user provides a bid price that amounts to a total cost that is less than or equal to their coin balance, then the transaction is successful and the buy order is processed.
+
+- GIVEN - user provides a bid price that amounts to a total cost that is less than or equal to their coin balance
+- WHEN - user tries to create a buy order
+- THEN - transaction is successful, buy order is processed
+
+If a user tries to create a buy order and the user provides a bid price that amounts to a total cost that is more than their coin balance, then the transaction fails and the buy order is NOT processed.
+
+- GIVEN - user provides a bid price that amounts to a total cost that is more than their coin balance
+- WHEN - user tries to create a buy order
+- THEN - transaction fails, buy order is NOT processed
+
+If a user tries to create a buy order and the user provides a bid price using an ask denom that is allowed, then the transaction is successful and the buy order is processed.
+
+- GIVEN - user provides a bid price using an ask denom that is allowed
+- WHEN - user tries to create a buy order
+- THEN - transaction is successful, buy order is processed
+
+If a user tries to create a buy order and the user provides a bid price using an ask denom that is NOT allowed, then the transaction fails and the buy order is NOT processed.
+
+- GIVEN - user provides a bid price using an ask denom that is NOT allowed
+- WHEN - user tries to create a buy order
+- THEN - transaction fails, buy order is NOT processed
+
+## Allow Ask Denom
+
+If a user tries to add an ask denom and the user submits a governance proposal that is approved, then the transaction is successful and the ask denom is added.
```

Comment:
do we need to link the _new ask denom_ message to a gov proposal? Or it is automatically handled by the gov proposal execution?

Reply:
The gov proposal executes the AllowAskDenom message. I can clarify.